### PR TITLE
Reduce API surface of tagged wrapped hash types

### PR DESCRIPTION
**Pull request by tcharding**
Recently we made it so that wrapper types created with `hash_newtype` were not general purpose hash types i.e., one could not easily hash arbitrary data into them. We would like to do the same for tagged wrapped hash types.

In `hashes` do:

- Create a new macro `sha256_tag` that does just the tag/engine stuff out of the `sha256t_hash_newtype` macro. 
- Deprecate the `sha256t_hash_newtype` macro.

In `bitcoin` do:

- Use a combination of `sha256_tag` and `hash_newtype` to create tagged wrapped hash types.

Note that we do not add private helper functions `engine` and `from_engine` to the tagged wrapper types as we do for legacy/segwit in `sighash`. Can be done later if wanted/needed.

Fix: #3135

### CODE REVIEW

**Comment by tcharding** [comment]
Good idea, can do.

**Comment by tcharding** [comment]
Done all in a single commit now, re-wrote PR title and description. 

FTR I'm a bit concerned that since we slightly switched gears (to be more conscious of deprectating not breaking) half way through a massive overhaul of `hashes` the `deprecated since` is going to imply that deprecated things can continue to be used as in the last release - with so many changes to `hashes` I'm not confident that that is the case.

File: hashes/src/sha256t.rs [hunk 1]
```diff
@@ -195,14 +240,15 @@ where
 ///
```
Comment (Kixunil) [inline hashes/src/sha256t.rs:240]: FTR this still claims support for multiple tags but it's not there anymore.

File: hashes/src/sha256t.rs [hunk 2]
```diff
@@ -266,7 +312,7 @@ macro_rules! sha256t_hash_newtype {
 // Workaround macros being unavailable in attributes.
```
Comment (Kixunil) [inline hashes/src/sha256t.rs:312]: BTW this is no longer needed in 1.63 and we can just move this to the inside of the other macro. (Maybe this was also available in 1.56, IDK.)

**Review by Kixunil — CHANGES_REQUESTED** [review]
The doc needs serious updates.

File: hashes/src/sha256t.rs (hunk 2 above)
Comment (tcharding) [reply hashes/src/sha256t.rs:293]: Will do as follow up.

**Comment by tcharding** [comment]
All docs suggestions implemented, apologies for the sloppy PR.

**Comment by apoelstra** [comment]
In the commit message and deprecation notice you say `sha256_tag` but the actual macro name is `sha256t_tag`.

**Comment by apoelstra** [comment]
Unrelated to this PR:

* We should really figure out how to make the `engine` method that produces the tagged hash engine constant. (But this is hard to do and fine to leave for now.)
* I wonder if we should make the `hash_newtype` macro produce a **private** method for converting from the underlying hashtype. So then we could use that instead of "casting" hashes, and get a bit of type safety.

**Comment by tcharding** [comment]
> I wonder if we should make the hash_newtype macro produce a private method for converting from the underlying hashtype. So then we could use that instead of "casting" hashes, and get a bit of type safety.

That sounds easy enough, I threw it on my todo list.

**Comment by tcharding** [comment]
Force push is just fix to commit log of the first patch.

**Comment by apoelstra** [comment]
It does not appear that you force-pushed.

**Comment by tcharding** [comment]
Woops

**Comment by Kixunil** [comment]
> * We should really figure out how to make the `engine` method that produces the tagged hash engine constant. (But this is hard to do and fine to leave for now.)

After these changes I think `Tag::engine()` is quite reasonable. We should make sure it's (also) an inherent method though.

> * I wonder if we should make the `hash_newtype` macro produce a **private** method for converting from the underlying hashtype. So then we could use that instead of "casting" hashes, and get a bit of type safety.

Oh, I thought this would be go against our goal of stabilizing primitives but I've just realized it actually doesn't as long as we don't expose it somewhere by reexporting the type. It'd also make sense to newtype the engine.

File: hashes/src/sha256t.rs [hunk 3]
```diff
@@ -220,15 +220,6 @@ macro_rules! sha256t_tag {
 /// BIP-341. Alternatively, you can supply `hash_bytes` to hash raw bytes. If you have the midstate
 /// already pre-computed and prefer **compiler** performance to readability you may use
 /// `raw(MIDSTATE_BYTES, HASHED_BYTES_LENGTH)` instead.
-///
-/// Both visibility modifiers and attributes are optional and passed to inner structs (excluding
-/// `#[hash_newtype(...)]`). The attributes suffer same compiler performance limitations as in
-/// [`hash_newtype`] macro.
```
Comment (Kixunil) [inline hashes/src/sha256t.rs:191]: The performance issue is still *here* it just wasn't in the `sha256t_tag` macro.

What causes performance issues for large inputs (quadratic algo) is our use of `tt` muncher which we still do use in `hash_newtype!` to process the atrributes.

**Review by Kixunil — CHANGES_REQUESTED** [review]

File: hashes/src/sha256t.rs (hunk 3 above)
Comment (tcharding) [reply hashes/src/sha256t.rs:191]: Oh, thanks for explaining.

**Comment by tcharding** [comment]
I put the perf docs back in.

**Review by Kixunil — APPROVED** [review]
ACK c97389596b9a18d3a9dcfbce3ade7274268483d2

**Review by apoelstra — APPROVED** [review]
ACK c97389596b9a18d3a9dcfbce3ade7274268483d2 successfully ran local tests

**Comment by apoelstra** [comment]
FYI "successfully ran local tests" today (and all last week) is running a degraded version of local tests due to #3195. This is why CI is now broken on master despite my running local CI on every merge commit.

**Pull request merged**